src/Microsoft.Data/PrimitiveColumn.cs

```diff
@@ -17,6 +19,47 @@ public partial class PrimitiveColumn<T> : BaseColumn
     {
         private PrimitiveColumnContainer<T> _columnContainer;
 
+        internal int NumberOfBuffers => _columnContainer.Buffers.Count;
+        internal Memory<byte> Buffer(int i) => _columnContainer.Buffers[i].Memory;
+        internal Memory<byte> NullBuffer(int i) => _columnContainer.NullBitMapBuffers[i].Memory;
+
+        internal Apache.Arrow.Array AsArrowArray(int i)
+        {
+            Debug.Assert(i < _columnContainer.Buffers.Count);
+            DataFrameBuffer<T> values = _columnContainer.Buffers[i];
+            DataFrameBuffer<byte> nulls = _columnContainer.NullBitMapBuffers[i];
+            ArrowBuffer valueBuffer = new ArrowBuffer(values.Memory);
+            // The following line is wrong, but C# Arrow does not have good null support yet
+            ArrowBuffer nullBuffer = nulls.Length == 0 ? ArrowBuffer.Empty : new ArrowBuffer(nulls.Memory);
+            switch (this)
+            {
+                case PrimitiveColumn<bool> boolColumn:
+                    return new BooleanArray(valueBuffer, nullBuffer, values.Length, -1, 0);
```

Comment:
Why `-1` for NullCount?

Reply:
Because we only store the NullCount for an entire column, we don't know the NullCount in each `DataFrameBuffer`. Similar to the behavior we get in Arrow when we `Slice`.(https://issues.apache.org/jira/browse/ARROW-5634)